File: 6-zigzag-conversion/6-zigzag-conversion.py
class Solution:
  def convert(self, s: str, numRows: int) -> str:
    if numRows == 1: return s
    result = ''
    n = len(s)
    step = 2*numRows - 2
    for i in range(numRows):
      index1 = i
      index2 = step - i if i % (numRows-1) else n
      while index1 < n or index2 < n:
        if index1 < n: 
          result += s[index1]
          index1 += step
        if index2 < n: 
          result += s[index2]
          index2 += step
    return result

File: 6-zigzag-conversion/6-zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
  let len = s.length
  if (numRows <= 1) return s
  let step = 2 * numRows - 2
  let result = ''
  for (let i = 0; i < numRows; i++) {
    let index = i;
    let index2 = i%(numRows-1) ? step - i : len;
    while (index < len || index2 < len) {
      if (index < len) {
        result += s[index]
        index += step
      }
      if (index2 < len) {
        result += s[index2]
        index2 += step
      }
    }
  }
  return result;
};